k9mail/src/main/java/com/fsck/k9/mailstore/migrations/MigrationTo58.java

```diff
@@ -0,0 +1,31 @@
+package com.fsck.k9.mailstore.migrations;
+
+
+import android.database.sqlite.SQLiteDatabase;
+
+
+class MigrationTo58 {
+    static void cleanUpOrphanedData(SQLiteDatabase db) {
+        cleanUpFtsTable(db);
+        cleanUpMessagePartsTable(db);
+    }
+    
+    private static void cleanUpFtsTable(SQLiteDatabase db) {
+        MigrationTo56.cleanUpFtsTable(db);
+    }
+
+    private static void cleanUpMessagePartsTable(SQLiteDatabase db) {
+        db.execSQL("DELETE FROM message_parts WHERE root NOT IN " +
+                "(SELECT message_part_id FROM messages WHERE deleted = 0)");
```

Comment:
@cketti 
In my database, there are rows with `deleted=0` and `message_part_id=NULL` (*). Therefore, the result of this subselect (line 19) contains `NULL` values. However, `row IN (NULL)` is always `true` and as a consequence, no message matches the negation and nothing is deleted.

Hence, this still doesn't fix the problem in all cases!

I propose to change the subselect to `(SELECT message_part_id FROM messages WHERE deleted = 0 AND message_part_id IS NOT NULL)`. But you have to decide if it is sufficient to change `MigrationTo58` (since the version wasn't released yet) or if a new database version is required (since some people may use the version from git).

(*) These entries with `deleted=0` and `message_part_id=NULL` only contain the `message_id` and some flags (e.g. `empty=1`). It seems that I had never a message with that message ID, but there was another message (with another message ID) which references to this message ID in its header fields `In-Reply-To` and `References` (e.g. GitHub notifications).